File: 295-find-median-from-data-stream/find-median-from-data-stream.cpp
class MedianFinder {
public:
    priority_queue<int, vector<int>, greater<int>> minHeap;
    priority_queue<int> maxHeap;
    MedianFinder() {
        
    }
    
    void addNum(int num) {
        maxHeap.push(num);
        while(maxHeap.size()>1+minHeap.size()){
            minHeap.push(maxHeap.top());
            maxHeap.pop();
        }
        while(!maxHeap.empty() && !minHeap.empty()){
            if(maxHeap.top()<=minHeap.top()) break;
            int val1= maxHeap.top(), val2= minHeap.top();
            maxHeap.pop(), minHeap.pop();
            maxHeap.push(val2);
            minHeap.push(val1);
        }
    }
    
    double findMedian() {
        if(minHeap.size()==maxHeap.size()){
            return (double)(maxHeap.top()+minHeap.top())/(double)2.0;
        }
        else return maxHeap.top();
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 295-find-median-from-data-stream/find-median-from-data-stream.java
class MedianFinder {
    PriorityQueue<Integer> maxHeap, minHeap;
    public MedianFinder() {
        minHeap= new PriorityQueue<>();
        maxHeap= new PriorityQueue<>((a,b)-> b-a);
    }
    
    public void addNum(int num) {
        if(minHeap.size()==maxHeap.size()){
            minHeap.offer(num);
            maxHeap.offer(minHeap.poll());
        }
        else{
            maxHeap.offer(num);
            minHeap.offer(maxHeap.poll());
        }
    }
    
    public double findMedian() {
        if(minHeap.size()==maxHeap.size()){
            return (minHeap.peek()+maxHeap.peek())/2.0;
        }
        else return maxHeap.peek()*1.0;
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */